notdiamond/toolkit/retry.py

```diff
@@ -0,0 +1,505 @@
+            else user_messages
+        )
+        return kwargs
+
+    def _get_fallback_model(self, failed_models: List[str] = []) -> str:
+        """
+        After failing to invoke current_model, use the user's model fallback list to choose the
+        next invocation model.
+        """
+        models = [m for m in self._models if m not in failed_models]
+
+        if not models:
+            return
+        return models[0]
+
+    def _retry_decorator(self, func):
+        @wraps(func)
+        async def async_wrapper(*args, **kwargs):
+            target_model = kwargs["model"]
+            base_messages = kwargs.get("messages", [])
+            kwargs = self._update_model_kwargs(
+                kwargs, target_model, user_messages=base_messages
+            )
+
+            attempt = 0
+            failed_models = []
+            while attempt < self.get_max_retries(target_model):
+                try:
+                    return await func(*args, **kwargs)
+                except Exception as e:
+                    LOGGER.warning(f"Attempt {attempt + 1} failed: {str(e)}")
+
+                    if attempt == self.get_max_retries(target_model) - 1:
+                        previous_model = target_model
+                        failed_models.append(previous_model)
+
+                        # If we cannot fallback, raise the exception
+                        # If we need to load balance, let the manager handle it
+                        target_model = None
+                        if not isinstance(self._models, dict):
+                            target_model = self._get_fallback_model(
+                                failed_models
+                            )
+                        if not target_model:
+                            raise _RetryWrapperException(failed_models, e)
+
+                        kwargs = self._update_model_kwargs(
+                            kwargs, target_model, user_messages=base_messages
+                        )
+                        LOGGER.info(
+                            f"Attempting fallback model {kwargs['model']}"
+                        )
+                        attempt = 0
+                        continue
+
+                attempt += 1
+                await asyncio.sleep(self._backoff**attempt)
+
+        @wraps(func)
+        def sync_wrapper(*args, **kwargs):
+            target_model = kwargs["model"]
+            base_messages = kwargs.get("messages", [])
+            kwargs = self._update_model_kwargs(
+                kwargs, target_model, user_messages=base_messages
+            )
+
+            attempt = 0
+            failed_models = []
+            while attempt < self.get_max_retries(target_model):
+                try:
+                    return func(*args, **kwargs)
+                except Exception as e:
+                    LOGGER.warning(f"Attempt {attempt + 1} failed: {str(e)}")
+
+                    if attempt == self.get_max_retries(target_model) - 1:
```

Comment:
Just to check my understanding. This fallback behaviour basically goes through the list of models the user defined sequentially or goes through the list of models by descending order of weight if load balancing is specified. However, with the `RetryManager`, if load balancing is specified, it will override this behaviour by resampling from the distribution without the failed models. Am I correct?

Reply:
Discussed offline - with fresh eyes I noticed this didn't work as we would've preferred, I've fixed and tested this to work as expected.